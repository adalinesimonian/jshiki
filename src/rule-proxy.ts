import RuleTree, { AccessPathArray, AccessPathPart } from './rule-tree'

function hasAccess(
  ruleTree: RuleTree,
  explicitAllow: boolean,
  propertyScope: AccessPathArray
) {
  const match = ruleTree.match(propertyScope)
  return explicitAllow
    ? match === true || (typeof match === 'string' && match === 'allow')
    : !(typeof match === 'string' && match === 'block')
}

function getProxyHandler<T extends object>(
  ruleTree: RuleTree,
  explicitAllow: boolean,
  scope: AccessPathArray = []
): ProxyHandler<T> {
  const cache = new Map<AccessPathPart, any>()

  return {
    get(target: T, property: AccessPathPart) {
      if (cache.has(property)) {
        return cache.get(property)
      }

      const propertyScope = scope.concat(property)

      if (!hasAccess(ruleTree, explicitAllow, propertyScope)) {
        cache.set(property, undefined)
        return undefined
      }

      const propValue = (target as any)[property]
      const returnValue =
        (propValue && typeof propValue === 'object') ||
        typeof propValue === 'function'
          ? new Proxy(
              (target as any)[property],
              getProxyHandler(ruleTree, explicitAllow, propertyScope)
            )
          : propValue
      cache.set(property, returnValue)
      return returnValue
    },

    set(_target: T, _property: AccessPathPart, value: any) {
      return value
    },
  }
}

export default function getRuleProxy<T extends object>(
  target: T,
  ruleTree: RuleTree,
  explicitAllow?: boolean
): T {
  return (target && typeof target === 'object') || typeof target === 'function'
    ? new Proxy(target, getProxyHandler(ruleTree, Boolean(explicitAllow)))
    : target
}
