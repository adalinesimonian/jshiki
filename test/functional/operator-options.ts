import { evaluate, evaluateAsync, parse, parseAsync } from '../../src'

describe('Operator options', () => {
  it('should allow all supported operators by default', () => {
    // Unary operators
    expect(evaluate('+2')).toBe(2)
    expect(evaluate('-2')).toBe(-2)
    expect(evaluate('!2')).toBe(false)
    expect(evaluate('~2')).toBe(-3)

    // Binary operators
    expect(evaluate('4 + 2')).toBe(6)
    expect(evaluate('4 - 2')).toBe(2)
    expect(evaluate('4 * 2')).toBe(8)
    expect(evaluate('4 ** 2')).toBe(16)
    expect(evaluate('4 / 2')).toBe(2)
    expect(evaluate('4 % 2')).toBe(0)
    expect(evaluate('4 < 2')).toBe(false)
    expect(evaluate('4 > 2')).toBe(true)
    expect(evaluate('4 <= 2')).toBe(false)
    expect(evaluate('4 >= 2')).toBe(true)
    expect(evaluate('4 == 2')).toBe(false)
    expect(evaluate('4 != 2')).toBe(true)
    expect(evaluate('4 === 2')).toBe(false)
    expect(evaluate('4 !== 2')).toBe(true)
    expect(evaluate('4 | 2')).toBe(6)
    expect(evaluate('4 ^ 2')).toBe(6)
    expect(evaluate('4 & 2')).toBe(0)
    expect(evaluate('4 << 2')).toBe(16)
    expect(evaluate('4 >> 2')).toBe(1)
    expect(evaluate('4 >>> 2')).toBe(1)

    // Logical operators
    expect(evaluate('null && true')).toBe(null)
    expect(evaluate('null || true')).toBe(true)
    expect(evaluate('null ?? true')).toBe(true)

    // Conditional operator
    expect(evaluate('false ? 5 : 3')).toBe(3)
  })

  it('should support allowing only the specified operators', () => {
    const options = {
      operators: {
        unary: { allow: ['+', '-'] },
        binary: { allow: ['%', '**'] },
        logical: { allow: ['&&'] },
        ternary: true,
      },
    }

    // Unary operators
    expect(evaluate('+2', options)).toBe(2)
    expect(evaluate('-2', options)).toBe(-2)
    expect(() => parse('!2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('~2', options)).toThrowErrorMatchingSnapshot()

    // Binary operators
    expect(() => parse('4 + 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 - 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 * 2', options)).toThrowErrorMatchingSnapshot()
    expect(evaluate('4 ** 2', options)).toBe(16)
    expect(() => parse('4 / 2', options)).toThrowErrorMatchingSnapshot()
    expect(evaluate('4 % 2', options)).toBe(0)
    expect(() => parse('4 < 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 > 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 <= 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 >= 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 == 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 != 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 === 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 !== 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 | 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 ^ 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 & 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 << 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 >> 2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('4 >>> 2', options)).toThrowErrorMatchingSnapshot()

    // Logical operators
    expect(evaluate('null && true', options)).toBe(null)
    expect(() => parse('null || true', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('null ?? true', options)).toThrowErrorMatchingSnapshot()

    // Conditional operator
    expect(evaluate('false ? 5 : 3', options)).toBe(3)
  })

  it('should support blocking specified operators', () => {
    const options = {
      operators: {
        unary: { block: ['+', '-'] },
        binary: { block: ['%', '**'] },
        logical: { block: ['&&'] },
        ternary: false,
      },
    }

    // Unary operators
    expect(() => parse('+2', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('-2', options)).toThrowErrorMatchingSnapshot()
    expect(evaluate('!2', options)).toBe(false)
    expect(evaluate('~2', options)).toBe(-3)

    // Binary operators
    expect(evaluate('4 + 2', options)).toBe(6)
    expect(evaluate('4 - 2', options)).toBe(2)
    expect(evaluate('4 * 2', options)).toBe(8)
    expect(() => parse('4 ** 2', options)).toThrowErrorMatchingSnapshot()
    expect(evaluate('4 / 2', options)).toBe(2)
    expect(() => parse('4 % 2', options)).toThrowErrorMatchingSnapshot()
    expect(evaluate('4 < 2', options)).toBe(false)
    expect(evaluate('4 > 2', options)).toBe(true)
    expect(evaluate('4 <= 2', options)).toBe(false)
    expect(evaluate('4 >= 2', options)).toBe(true)
    expect(evaluate('4 == 2', options)).toBe(false)
    expect(evaluate('4 != 2', options)).toBe(true)
    expect(evaluate('4 === 2', options)).toBe(false)
    expect(evaluate('4 !== 2', options)).toBe(true)
    expect(evaluate('4 | 2', options)).toBe(6)
    expect(evaluate('4 ^ 2', options)).toBe(6)
    expect(evaluate('4 & 2', options)).toBe(0)
    expect(evaluate('4 << 2', options)).toBe(16)
    expect(evaluate('4 >> 2', options)).toBe(1)
    expect(evaluate('4 >>> 2', options)).toBe(1)

    // Logical operators
    expect(() => parse('null && true', options)).toThrowErrorMatchingSnapshot()
    expect(evaluate('null || true', options)).toBe(true)
    expect(evaluate('null ?? true', options)).toBe(true)

    // Conditional operator
    expect(() => parse('false ? 5 : 3', options)).toThrowErrorMatchingSnapshot()
  })

  it('should throw if providing both an allow and block list', () => {
    expect(() =>
      parse('1 + 2', {
        operators: {
          unary: { allow: ['+', '-'], block: ['+', '-'] },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          binary: { allow: ['+', '-'], block: ['+', '-'] },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          logical: { allow: ['&&', '??'], block: ['&&', '??'] },
        },
      })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should throw if providing an unsupported operator', () => {
    expect(() =>
      parse('1 + 2', {
        operators: {
          unary: { allow: ['invalid'] },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          unary: { block: ['invalid'] },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          binary: { allow: ['invalid'] },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          binary: { block: ['invalid'] },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          logical: { allow: ['invalid'] },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          logical: { block: ['invalid'] },
        },
      })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should throw if the list is not an array', () => {
    expect(() =>
      parse('1 + 2', {
        operators: {
          unary: { allow: 'invalid' as any },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          unary: { block: 'invalid' as any },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          binary: { allow: 'invalid' as any },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          binary: { block: 'invalid' as any },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          logical: { allow: 'invalid' as any },
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          logical: { block: 'invalid' as any },
        },
      })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should throw if the options do not contain a list', () => {
    expect(() =>
      parse('1 + 2', {
        operators: {
          unary: {} as any,
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          unary: {} as any,
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          binary: {} as any,
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          binary: {} as any,
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          logical: {} as any,
        },
      })
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('1 + 2', {
        operators: {
          logical: {} as any,
        },
      })
    ).toThrowErrorMatchingSnapshot()
  })
})

describe('Operator options (async)', () => {
  it('should allow all supported operators by default', async () => {
    // Unary operators
    expect(await evaluateAsync('+2')).toBe(2)
    expect(await evaluateAsync('-2')).toBe(-2)
    expect(await evaluateAsync('!2')).toBe(false)
    expect(await evaluateAsync('~2')).toBe(-3)

    // Binary operators
    expect(await evaluateAsync('4 + 2')).toBe(6)
    expect(await evaluateAsync('4 - 2')).toBe(2)
    expect(await evaluateAsync('4 * 2')).toBe(8)
    expect(await evaluateAsync('4 ** 2')).toBe(16)
    expect(await evaluateAsync('4 / 2')).toBe(2)
    expect(await evaluateAsync('4 % 2')).toBe(0)
    expect(await evaluateAsync('4 < 2')).toBe(false)
    expect(await evaluateAsync('4 > 2')).toBe(true)
    expect(await evaluateAsync('4 <= 2')).toBe(false)
    expect(await evaluateAsync('4 >= 2')).toBe(true)
    expect(await evaluateAsync('4 == 2')).toBe(false)
    expect(await evaluateAsync('4 != 2')).toBe(true)
    expect(await evaluateAsync('4 === 2')).toBe(false)
    expect(await evaluateAsync('4 !== 2')).toBe(true)
    expect(await evaluateAsync('4 | 2')).toBe(6)
    expect(await evaluateAsync('4 ^ 2')).toBe(6)
    expect(await evaluateAsync('4 & 2')).toBe(0)
    expect(await evaluateAsync('4 << 2')).toBe(16)
    expect(await evaluateAsync('4 >> 2')).toBe(1)
    expect(await evaluateAsync('4 >>> 2')).toBe(1)

    // Logical operators
    expect(await evaluateAsync('null && true')).toBe(null)
    expect(await evaluateAsync('null || true')).toBe(true)
    expect(await evaluateAsync('null ?? true')).toBe(true)

    // Conditional operator
    expect(await evaluateAsync('false ? 5 : 3')).toBe(3)
  })

  it('should support allowing only the specified operators', async () => {
    const options = {
      operators: {
        unary: { allow: ['+', '-'] },
        binary: { allow: ['%', '**'] },
        logical: { allow: ['&&'] },
        ternary: true,
      },
    }

    // Unary operators
    expect(await evaluateAsync('+2', options)).toBe(2)
    expect(await evaluateAsync('-2', options)).toBe(-2)
    expect(async () =>
      parseAsync('!2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('~2', options)
    ).rejects.toThrowErrorMatchingSnapshot()

    // Binary operators
    expect(async () =>
      parseAsync('4 + 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 - 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 * 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(await evaluateAsync('4 ** 2', options)).toBe(16)
    expect(async () =>
      parseAsync('4 / 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(await evaluateAsync('4 % 2', options)).toBe(0)
    expect(async () =>
      parseAsync('4 < 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 > 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 <= 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 >= 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 == 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 != 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 === 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 !== 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 | 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 ^ 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 & 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 << 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 >> 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('4 >>> 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()

    // Logical operators
    expect(await evaluateAsync('null && true', options)).toBe(null)
    expect(async () =>
      parseAsync('null || true', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('null ?? true', options)
    ).rejects.toThrowErrorMatchingSnapshot()

    // Conditional operator
    expect(await evaluateAsync('false ? 5 : 3', options)).toBe(3)
  })

  it('should support blocking specified operators', async () => {
    const options = {
      operators: {
        unary: { block: ['+', '-'] },
        binary: { block: ['%', '**'] },
        logical: { block: ['&&'] },
        ternary: false,
      },
    }

    // Unary operators
    expect(async () =>
      parseAsync('+2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('-2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(await evaluateAsync('!2', options)).toBe(false)
    expect(await evaluateAsync('~2', options)).toBe(-3)

    // Binary operators
    expect(await evaluateAsync('4 + 2', options)).toBe(6)
    expect(await evaluateAsync('4 - 2', options)).toBe(2)
    expect(await evaluateAsync('4 * 2', options)).toBe(8)
    expect(async () =>
      parseAsync('4 ** 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(await evaluateAsync('4 / 2', options)).toBe(2)
    expect(async () =>
      parseAsync('4 % 2', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(await evaluateAsync('4 < 2', options)).toBe(false)
    expect(await evaluateAsync('4 > 2', options)).toBe(true)
    expect(await evaluateAsync('4 <= 2', options)).toBe(false)
    expect(await evaluateAsync('4 >= 2', options)).toBe(true)
    expect(await evaluateAsync('4 == 2', options)).toBe(false)
    expect(await evaluateAsync('4 != 2', options)).toBe(true)
    expect(await evaluateAsync('4 === 2', options)).toBe(false)
    expect(await evaluateAsync('4 !== 2', options)).toBe(true)
    expect(await evaluateAsync('4 | 2', options)).toBe(6)
    expect(await evaluateAsync('4 ^ 2', options)).toBe(6)
    expect(await evaluateAsync('4 & 2', options)).toBe(0)
    expect(await evaluateAsync('4 << 2', options)).toBe(16)
    expect(await evaluateAsync('4 >> 2', options)).toBe(1)
    expect(await evaluateAsync('4 >>> 2', options)).toBe(1)

    // Logical operators
    expect(async () =>
      parseAsync('null && true', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(await evaluateAsync('null || true', options)).toBe(true)
    expect(await evaluateAsync('null ?? true', options)).toBe(true)

    // Conditional operator
    expect(async () =>
      parseAsync('false ? 5 : 3', options)
    ).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should throw if providing both an allow and block list', async () => {
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          unary: { allow: ['+', '-'], block: ['+', '-'] },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          binary: { allow: ['+', '-'], block: ['+', '-'] },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          logical: { allow: ['&&', '??'], block: ['&&', '??'] },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should throw if providing an unsupported operator', async () => {
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          unary: { allow: ['invalid'] },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          unary: { block: ['invalid'] },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          binary: { allow: ['invalid'] },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          binary: { block: ['invalid'] },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          logical: { allow: ['invalid'] },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          logical: { block: ['invalid'] },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should throw if the list is not an array', async () => {
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          unary: { allow: 'invalid' as any },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          unary: { block: 'invalid' as any },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          binary: { allow: 'invalid' as any },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          binary: { block: 'invalid' as any },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          logical: { allow: 'invalid' as any },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          logical: { block: 'invalid' as any },
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should throw if the options do not contain a list', async () => {
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          unary: {} as any,
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          unary: {} as any,
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          binary: {} as any,
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          binary: {} as any,
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          logical: {} as any,
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('1 + 2', {
        operators: {
          logical: {} as any,
        },
      })
    ).rejects.toThrowErrorMatchingSnapshot()
  })
})
