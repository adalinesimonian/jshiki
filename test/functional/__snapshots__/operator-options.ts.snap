// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operator options (async) should support allowing only the specified operators 1`] = `"Unary operator ! is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 2`] = `"Unary operator ~ is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 3`] = `"Binary operator + is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 4`] = `"Binary operator - is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 5`] = `"Binary operator * is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 6`] = `"Binary operator / is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 7`] = `"Binary operator < is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 8`] = `"Binary operator > is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 9`] = `"Binary operator <= is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 10`] = `"Binary operator >= is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 11`] = `"Binary operator == is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 12`] = `"Binary operator != is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 13`] = `"Binary operator === is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 14`] = `"Binary operator !== is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 15`] = `"Binary operator | is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 16`] = `"Binary operator ^ is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 17`] = `"Binary operator & is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 18`] = `"Binary operator << is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 19`] = `"Binary operator >> is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 20`] = `"Binary operator >>> is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 21`] = `"Logical operator || is not allowed"`;

exports[`Operator options (async) should support allowing only the specified operators 22`] = `"Logical operator ?? is not allowed"`;

exports[`Operator options (async) should support blocking specified operators 1`] = `"Unary operator + is not allowed"`;

exports[`Operator options (async) should support blocking specified operators 2`] = `"Unary operator - is not allowed"`;

exports[`Operator options (async) should support blocking specified operators 3`] = `"Binary operator ** is not allowed"`;

exports[`Operator options (async) should support blocking specified operators 4`] = `"Binary operator % is not allowed"`;

exports[`Operator options (async) should support blocking specified operators 5`] = `"Logical operator && is not allowed"`;

exports[`Operator options (async) should support blocking specified operators 6`] = `"Conditional/ternary operator is not allowed"`;

exports[`Operator options (async) should throw if providing an unsupported operator 1`] = `"Operator invalid is not supported"`;

exports[`Operator options (async) should throw if providing an unsupported operator 2`] = `"Operator invalid is not supported"`;

exports[`Operator options (async) should throw if providing an unsupported operator 3`] = `"Operator invalid is not supported"`;

exports[`Operator options (async) should throw if providing an unsupported operator 4`] = `"Operator invalid is not supported"`;

exports[`Operator options (async) should throw if providing an unsupported operator 5`] = `"Operator invalid is not supported"`;

exports[`Operator options (async) should throw if providing an unsupported operator 6`] = `"Operator invalid is not supported"`;

exports[`Operator options (async) should throw if providing both an allow and block list 1`] = `"Cannot specify both an allow list and block list of operators"`;

exports[`Operator options (async) should throw if providing both an allow and block list 2`] = `"Cannot specify both an allow list and block list of operators"`;

exports[`Operator options (async) should throw if providing both an allow and block list 3`] = `"Cannot specify both an allow list and block list of operators"`;

exports[`Operator options (async) should throw if the list is not an array 1`] = `"Operator allow list must be an array"`;

exports[`Operator options (async) should throw if the list is not an array 2`] = `"Operator block list must be an array"`;

exports[`Operator options (async) should throw if the list is not an array 3`] = `"Operator allow list must be an array"`;

exports[`Operator options (async) should throw if the list is not an array 4`] = `"Operator block list must be an array"`;

exports[`Operator options (async) should throw if the list is not an array 5`] = `"Operator allow list must be an array"`;

exports[`Operator options (async) should throw if the list is not an array 6`] = `"Operator block list must be an array"`;

exports[`Operator options (async) should throw if the options do not contain a list 1`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options (async) should throw if the options do not contain a list 2`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options (async) should throw if the options do not contain a list 3`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options (async) should throw if the options do not contain a list 4`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options (async) should throw if the options do not contain a list 5`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options (async) should throw if the options do not contain a list 6`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options should support allowing only the specified operators 1`] = `"Unary operator ! is not allowed"`;

exports[`Operator options should support allowing only the specified operators 2`] = `"Unary operator ~ is not allowed"`;

exports[`Operator options should support allowing only the specified operators 3`] = `"Binary operator + is not allowed"`;

exports[`Operator options should support allowing only the specified operators 4`] = `"Binary operator - is not allowed"`;

exports[`Operator options should support allowing only the specified operators 5`] = `"Binary operator * is not allowed"`;

exports[`Operator options should support allowing only the specified operators 6`] = `"Binary operator / is not allowed"`;

exports[`Operator options should support allowing only the specified operators 7`] = `"Binary operator < is not allowed"`;

exports[`Operator options should support allowing only the specified operators 8`] = `"Binary operator > is not allowed"`;

exports[`Operator options should support allowing only the specified operators 9`] = `"Binary operator <= is not allowed"`;

exports[`Operator options should support allowing only the specified operators 10`] = `"Binary operator >= is not allowed"`;

exports[`Operator options should support allowing only the specified operators 11`] = `"Binary operator == is not allowed"`;

exports[`Operator options should support allowing only the specified operators 12`] = `"Binary operator != is not allowed"`;

exports[`Operator options should support allowing only the specified operators 13`] = `"Binary operator === is not allowed"`;

exports[`Operator options should support allowing only the specified operators 14`] = `"Binary operator !== is not allowed"`;

exports[`Operator options should support allowing only the specified operators 15`] = `"Binary operator | is not allowed"`;

exports[`Operator options should support allowing only the specified operators 16`] = `"Binary operator ^ is not allowed"`;

exports[`Operator options should support allowing only the specified operators 17`] = `"Binary operator & is not allowed"`;

exports[`Operator options should support allowing only the specified operators 18`] = `"Binary operator << is not allowed"`;

exports[`Operator options should support allowing only the specified operators 19`] = `"Binary operator >> is not allowed"`;

exports[`Operator options should support allowing only the specified operators 20`] = `"Binary operator >>> is not allowed"`;

exports[`Operator options should support allowing only the specified operators 21`] = `"Logical operator || is not allowed"`;

exports[`Operator options should support allowing only the specified operators 22`] = `"Logical operator ?? is not allowed"`;

exports[`Operator options should support blocking specified operators 1`] = `"Unary operator + is not allowed"`;

exports[`Operator options should support blocking specified operators 2`] = `"Unary operator - is not allowed"`;

exports[`Operator options should support blocking specified operators 3`] = `"Binary operator ** is not allowed"`;

exports[`Operator options should support blocking specified operators 4`] = `"Binary operator % is not allowed"`;

exports[`Operator options should support blocking specified operators 5`] = `"Logical operator && is not allowed"`;

exports[`Operator options should support blocking specified operators 6`] = `"Conditional/ternary operator is not allowed"`;

exports[`Operator options should throw if providing an unsupported operator 1`] = `"Operator invalid is not supported"`;

exports[`Operator options should throw if providing an unsupported operator 2`] = `"Operator invalid is not supported"`;

exports[`Operator options should throw if providing an unsupported operator 3`] = `"Operator invalid is not supported"`;

exports[`Operator options should throw if providing an unsupported operator 4`] = `"Operator invalid is not supported"`;

exports[`Operator options should throw if providing an unsupported operator 5`] = `"Operator invalid is not supported"`;

exports[`Operator options should throw if providing an unsupported operator 6`] = `"Operator invalid is not supported"`;

exports[`Operator options should throw if providing both an allow and block list 1`] = `"Cannot specify both an allow list and block list of operators"`;

exports[`Operator options should throw if providing both an allow and block list 2`] = `"Cannot specify both an allow list and block list of operators"`;

exports[`Operator options should throw if providing both an allow and block list 3`] = `"Cannot specify both an allow list and block list of operators"`;

exports[`Operator options should throw if the list is not an array 1`] = `"Operator allow list must be an array"`;

exports[`Operator options should throw if the list is not an array 2`] = `"Operator block list must be an array"`;

exports[`Operator options should throw if the list is not an array 3`] = `"Operator allow list must be an array"`;

exports[`Operator options should throw if the list is not an array 4`] = `"Operator block list must be an array"`;

exports[`Operator options should throw if the list is not an array 5`] = `"Operator allow list must be an array"`;

exports[`Operator options should throw if the list is not an array 6`] = `"Operator block list must be an array"`;

exports[`Operator options should throw if the options do not contain a list 1`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options should throw if the options do not contain a list 2`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options should throw if the options do not contain a list 3`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options should throw if the options do not contain a list 4`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options should throw if the options do not contain a list 5`] = `"Operator options must specify either an allow list or a block list"`;

exports[`Operator options should throw if the options do not contain a list 6`] = `"Operator options must specify either an allow list or a block list"`;
